File: src/se/redmind/rmtest/selenium/grid/DriverProvider.java
package se.redmind.rmtest.selenium.grid;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.List;

import org.openqa.grid.common.RegistrationRequest;
import org.openqa.selenium.Platform;
import org.openqa.selenium.WebDriverException;
import org.openqa.selenium.remote.DesiredCapabilities;
import org.openqa.selenium.remote.SessionNotFoundException;

import se.redmind.rmtest.selenium.framework.Browser;
import se.redmind.rmtest.selenium.framework.DeviceDescription;
import se.redmind.rmtest.selenium.framework.config.FrameworkConfig;

/**
 * @author petter
 *
 */
public class DriverProvider {
	
	private static ArrayList <DriverNamingWrapper> urlCapList = new ArrayList<DriverNamingWrapper>();
	private static ArrayList <DriverNamingWrapper> allDrivers = new ArrayList<DriverNamingWrapper>();
	private static DesiredCapabilities currentCapability;
	
	
	/**
	 * 
	 */
	private static void updateDrivers() {
		final FrameworkConfig config = FrameworkConfig.getConfig();
		urlCapList = new ArrayList<DriverNamingWrapper>();
		if (!config.runOnGrid()) {
			loadLocalDrivers();
			return;
		}
		HubNodesStatus nodeInfo = new HubNodesStatus(config.getHubIp(), GridConstatants.hubPort);
		ArrayList <RegistrationRequest> nodeList = nodeInfo.getNodesAsRegReqs();

		RegistrationRequest nodeReq;
		String description;
		
		DescriptionBuilder descriptionBuilder = new DescriptionBuilder();
		for (int j = 0; j < nodeList.size(); j++) { 
			nodeReq = nodeList.get(j); 
			for (int i = 0; i < nodeReq.getCapabilities().size(); i++) {
				currentCapability = new DesiredCapabilities(nodeReq.getCapabilities().get(i));
				description = descriptionBuilder.buildDescriptionFromCapabilities(currentCapability);
				URL driverUrl;
				try {
					driverUrl = new URL("http://" + nodeReq.getConfigAsString("host") + ":" + nodeReq.getConfigAsString("port") + "/wd/hub");
					DriverNamingWrapper driver = new DriverNamingWrapper(driverUrl, currentCapability, description);
					urlCapList.add(driver);
					allDrivers.add(driver);
				} catch (MalformedURLException e) {
					// TODO Auto-generated catch block
					e.printStackTrace();
				}

			}

		}
	}


	private static void loadLocalDrivers() {
		final FrameworkConfig config = FrameworkConfig.getConfig();
		for (int i = 0; i < Browser.values().length; i++) {
			Browser browser = Browser.values()[i];
			if (browser == Browser.PhantomJS 	&& !config.usePhantomJS()) 	continue;
			if (browser == Browser.Chrome 		&& !config.useChrome()) 		continue;
			DriverNamingWrapper driver = new DriverNamingWrapper(browser, browser.toString());
			urlCapList.add(driver);
			allDrivers.add(driver);
		}
	}

	

	/**
	 * 
	 */
	public static void stopDrivers() {

		for (int i = 0; i < allDrivers.size(); i++) {
			System.out.println("Closing driver: " + allDrivers.get(i).getDescription());
			try {
				if (allDrivers.get(i).getDriver() != null) {
					allDrivers.get(i).getDriver().quit();
				}
				
			} catch (SessionNotFoundException e) {
				System.out.println("For some reason a session was gone while quitting");
				System.out.println(e);
				continue;
			}
			catch (WebDriverException e){
				System.out.println("Crached webdriver, continue to closing drivers");
				continue;
			}
		}
		allDrivers = new ArrayList<DriverNamingWrapper>();
	}

	
	/**
	 * 
	 * @return 
	 */
	public synchronized static Object[] getDrivers() {
		updateDrivers();
		
//		startDrivers(urlCapList);
		return urlCapList.toArray();
	}

	
	/**
	 * @param pPlatform
	 * @return
	 */
	public synchronized static Object[] getDrivers(Platform pPlatform) {
		updateDrivers();
		ArrayList <DriverNamingWrapper> filteredUrlCapList = new ArrayList<DriverNamingWrapper>();
		for (int i = 0; i < urlCapList.size(); i++) {
			if (urlCapList.get(i).getCapability().getPlatform().is(pPlatform)) {
				filteredUrlCapList.add(urlCapList.get(i));
			}
		}
		return filteredUrlCapList.toArray();
	}


	/**
	 * @param pPlatform1
	 * @param pPlatform2
	 * @return
	 */
	public synchronized static Object[] getDrivers(Platform pPlatform1, Platform pPlatform2) {
		updateDrivers();
		ArrayList <DriverNamingWrapper> filteredUrlCapList = new ArrayList<DriverNamingWrapper>();
		for (int i = 0; i < urlCapList.size(); i++) {
			if (urlCapList.get(i).getCapability().getPlatform().is(pPlatform1)) {
				filteredUrlCapList.add(urlCapList.get(i));
			}
			if (urlCapList.get(i).getCapability().getPlatform().is(pPlatform2)) {
				filteredUrlCapList.add(urlCapList.get(i));
			}
		}
		return filteredUrlCapList.toArray();
	}


	/**
	 * @param pPlatform
	 * @param pBrowserName
	 * @return
	 */
	public synchronized static Object[] getDrivers(Platform pPlatform, String pBrowserName) {
		updateDrivers();
		ArrayList <DriverNamingWrapper> filteredUrlCapList = new ArrayList<DriverNamingWrapper>();
		for (int i = 0; i < urlCapList.size(); i++) {
			if (urlCapList.get(i).getCapability().getPlatform().is(pPlatform)) {
				if (urlCapList.get(i).getCapability().getBrowserName().contains(pBrowserName)) {
					filteredUrlCapList.add(urlCapList.get(i));
				}
			}
		}		
		return filteredUrlCapList.toArray();
	}


	/**
	 * get drivers that match capability key,value pair
	 * @param capKey
	 * @param capValue
	 * @return
	 */
	public synchronized static Object[] getDrivers(String capKey, String capValue) {
		updateDrivers();
		ArrayList <DriverNamingWrapper> filteredUrlCapList = new ArrayList<DriverNamingWrapper>();
		String currCap;
		for (int i = 0; i < urlCapList.size(); i++) {
			currCap = (String) urlCapList.get(i).getCapability().getCapability(capKey);
			if (currCap == null) {
				currCap = "";
			}
			if (currCap.equalsIgnoreCase(capValue)) {
				filteredUrlCapList.add(urlCapList.get(i));
			}
		}
		return filteredUrlCapList.toArray();
	}
}

File: dotNet/RMTest/RMTest/DriverProvider.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Collections;
using Newtonsoft.Json.Linq;
using OpenQA.Selenium;
using OpenQA.Selenium.Remote;

namespace RMTest
{
	public class DriverProvider 
	{
		private static List<DriverNamingWrapper> driverList = new List<DriverNamingWrapper>();
		private static Boolean isInitialized = false;
    
		public static void startDrivers()
		{
			RmConfig config = new RmConfig();
			String hubHost = config.getHubIp();
			int hubPort = 4444;
			//HubNodesStatus nodeInfo = new HubNodesStatus(config.getHubIp(), GridConstatants.hubPort);
			HubNodesStatus nodeInfo = new HubNodesStatus(hubHost, hubPort);
			JArray nodeList = nodeInfo.getNodesAsJson();

			if (isInitialized) 
			{
				Console.WriteLine("Already started drivers.");
			}
			else 
			{
				//JObject nodeReq;
				String description;
				JArray capabilities;
				JObject configuration;
				foreach (JObject nodeReq in nodeList.Children<JObject>())
				{
					configuration = JObject.Parse(nodeReq.GetValue("configuration").ToString());
					capabilities = JArray.Parse(nodeReq.GetValue("capabilities").ToString());
					DesiredCapabilities capability;
					foreach (JObject jsonCapablility in capabilities.Children<JObject>())
					{
					//for (int i = 0; i < nodeReq.getCapabilities().size(); i++) {
						capability = new DesiredCapabilities();
						foreach (JProperty p in jsonCapablility.Properties())
						{
							  capability.SetCapability(p.Name, p.Value.ToString());  
						}
	//                    System.out.println("DESCRIPTION OF DEVICE: " + capability.getCapability("description") + nodeReq.getDescription());
					
						if (capability.GetCapability("description") != null) 
						{ 
                    		description = (String) capability.GetCapability("description");
						} 
						else if (nodeReq.GetValue("description") != null) 
						{
                    		description = nodeReq.GetValue("description").ToString();
						} 
						else 
						{
                    		description = capability.Platform + " " + capability.BrowserName;

						}
                    
						Console.WriteLine("Description of driver is: " + description);
                    


						try {
							Uri driverUrl = new Uri(configuration.GetValue("remoteHost") + "/wd/hub");
							IWebDriver driver;
							if (capability.GetCapability("app-package") == null) {
								driver = new RemoteWebDriver(driverUrl, capability);
							} else {
								driver = new RemoteWebDriver(driverUrl, capability);
								//driver = new SwipeableWebDriver(driverUrl, capability);
							}
                        
							driverList.Add(new DriverNamingWrapper(description, driver, capability, driverUrl));
							Console.WriteLine("Started driver: " + description);
              
    
						} catch (Exception e) {
							// TODO: handle exception
							Console.WriteLine(e.Message);
						}
                    
					}
				}
				isInitialized = true;
			}
		}
    
		//public static void cleanStaleDrivers() {
		//	for (int driverNr = 0; driverNr < driverList.size(); driverNr++) {
		//		try {
		//			System.out.println(driverList.get(driverNr).toString());
		//			driverList.get(driverNr).getDriver();
		//		} catch (Exception e) {
		//			// TODO: handle exception
		//			e.printStackTrace();
		//		}

		//	}
		//}

		public static void stopDrivers()
		{
			//DriverNamingWrapper DriveWrap;
			for (int i = 0; i < driverList.Count; i++)
			{
				//DriverNamingWrapper DriveWrap = driverList[i];
				Console.WriteLine("Closing driver: " + driverList[i].getDriverDescription());
				driverList[i].getDriver().Quit();

			}
		}

		/**
		 * 
		 * @return 
		 */
		public static void getDrivers() {
			startDrivers();
			for (int i = 0; i < driverList.Count; i++)
			{
				Console.WriteLine(driverList[i].toString());
			}
		
		}
    
	//	/**
	//	 * 
	//	 * @return 
	//	 */
	//	public synchronized static Object[] getDrivers(Platform pPlatform) {
	//		startDrivers();
	//		ArrayList <DriverNamingWrapper> filteredDriverList = new ArrayList<DriverNamingWrapper>();
	//		for (int i = 0; i < driverList.size(); i++) {
	//			if (driverList.get(i).getCapabilities().getPlatform().is(pPlatform)) {
	//				filteredDriverList.add(driverList.get(i));
	//			}
	//		}
	//		return filteredDriverList.toArray();
	//	}
    
	//	/**
	//	 * 
	//	 * @return 
	//	 */
	//	public synchronized static Object[] getDrivers(Platform pPlatform1, Platform pPlatform2) {
	//		startDrivers();
	//		ArrayList <DriverNamingWrapper> filteredDriverList = new ArrayList<DriverNamingWrapper>();
	//		for (int i = 0; i < driverList.size(); i++) {
	//			if (driverList.get(i).getCapabilities().getPlatform().is(pPlatform1)) {
	//				filteredDriverList.add(driverList.get(i));
	//			}
	//			if (driverList.get(i).getCapabilities().getPlatform().is(pPlatform2)) {
	//				filteredDriverList.add(driverList.get(i));
	//			}
	//		}
	//		return filteredDriverList.toArray();
	//	}
    
	//	/**
	//	 * 
	//	 * @return 
	//	 */
	//	public synchronized static Object[] getDrivers(Platform pPlatform, String pBrowserName) {
	//		startDrivers();
	//		ArrayList <DriverNamingWrapper> filteredDriverList = new ArrayList<DriverNamingWrapper>();
	//		for (int i = 0; i < driverList.size(); i++) {
	//			if (driverList.get(i).getCapabilities().getPlatform().is(pPlatform)) {
	//				if (driverList.get(i).getCapabilities().getBrowserName().contains(pBrowserName)) {
	////                    System.out.println("Choosen driver: " + driverList.get(i));
	//					filteredDriverList.add(driverList.get(i));
	//				}
	//			}
	//		}
	////        System.out.println("Number of drivers: " + filteredDriverList.size());
	//		return filteredDriverList.toArray();
	//	}
    
  
	//	/**
	//	 * get drivers that match capability key,value pair
	//	 * @param capKey
	//	 * @param capValue
	//	 * @return
	//	 */
	//	public synchronized static Object[] getDrivers(String capKey, String capValue) {
	//		startDrivers();
	//		ArrayList <DriverNamingWrapper> filteredDriverList = new ArrayList<DriverNamingWrapper>();
	//		String currCap;
	//		for (int i = 0; i < driverList.size(); i++) {
	//			currCap = (String) driverList.get(i).getCapabilities().getCapability(capKey);
	//			if (currCap == null) {
	//				currCap = "";
	//			}
	//			if (currCap.equalsIgnoreCase(capValue)) {
	//				filteredDriverList.add(driverList.get(i));
	//			}
	//		}
	//		return filteredDriverList.toArray();
	//	}
    
    
	}
}
